File: scripts/replay_clipboard_text.py
#!/usr/bin/env python3
"""Replay Hades successful text clipboard parity in an isolated direct PTY."""

from __future__ import annotations

import argparse
import errno
import fcntl
import json
import os
import pty
import re
import select
import shutil
import signal
import struct
import sys
import tempfile
import termios
import time
from pathlib import Path
from typing import Any, Callable


ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CONTRACT = ROOT / "tests/fixtures/parity/OBS-0023-hades-text-clipboard.json"
DEFAULT_BINARY = ROOT / "target/debug/hades"
ANSI_SEQUENCE = re.compile(r"\x1b(?:\[[0-?]*[ -/]*[@-~]|\][^\x07]*(?:\x07|\x1b\\)|[()][0-2A-Z])")
COLUMNS = 120
ROWS = 40


class ClipboardReplayFailure(RuntimeError):
    def __init__(self, case: str, step: str, message: str, details: dict[str, Any] | None = None):
        super().__init__(message)
        self.case = case
        self.step = step
        self.message = message
        self.details = details or {}

    def as_dict(self) -> dict[str, Any]:
        return {"case": self.case, "step": self.step, "message": self.message, **self.details}


def emit_report(report: dict[str, Any], path: Path | None) -> None:
    serialized = json.dumps(report, ensure_ascii=False, indent=2) + "\n"
    if path is not None:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(serialized, encoding="utf-8")
    print(serialized, end="")


def normalize(raw: bytes) -> str:
    text = raw.decode("utf-8", errors="replace").replace("\r", "")
    return ANSI_SEQUENCE.sub("", text).replace("\x08", "")


def contains_marker(raw: bytes, marker: str) -> bool:
    text = normalize(raw)
    compact_text = "".join(text.split()).lower()
    compact_marker = "".join(marker.split()).lower()
    return marker in text or compact_marker in compact_text


def set_window_size(fd: int) -> None:
    fcntl.ioctl(fd, termios.TIOCSWINSZ, struct.pack("HHHH", ROWS, COLUMNS, 0, 0))


def read_available(fd: int) -> bytes:
    chunks: list[bytes] = []
    while True:
        try:
            chunk = os.read(fd, 65_536)
        except OSError as error:
            if error.errno in (errno.EIO, errno.EAGAIN):
                break
            raise
        if not chunk:
            break
        chunks.append(chunk)
    return b"".join(chunks)


def child_status(pid: int) -> tuple[bool, int | None]:
    try:
        waited_pid, status = os.waitpid(pid, os.WNOHANG)
    except ChildProcessError:
        return True, None
    if waited_pid == 0:
        return False, None
    return True, status


def wait_for(
    pid: int,
    fd: int,
    buffer: bytes,
    case: str,
    step: str,
    predicate: Callable[[bytes], bool],
    timeout: float,
) -> bytes:
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if predicate(buffer):
            return buffer
        exited, status = child_status(pid)
        if exited:
            buffer += read_available(fd)
            if predicate(buffer):
                return buffer
            raise ClipboardReplayFailure(
                case,
                step,
                "process exited before the PTY assertion",
                {"exit_status": status, "screen_tail": normalize(buffer)[-2000:]},
            )
        readable, _, _ = select.select([fd], [], [], min(0.05, max(0.0, deadline - time.monotonic())))
        if readable:
            buffer += read_available(fd)
    raise ClipboardReplayFailure(
        case,
        step,
        f"timed out after {timeout:.1f}s",
        {"screen_tail": normalize(buffer)[-2000:]},
    )


def drain(pid: int, fd: int, buffer: bytes, duration: float = 0.1) -> bytes:
    deadline = time.monotonic() + duration
    while time.monotonic() < deadline:
        exited, _ = child_status(pid)
        if exited:
            return buffer + read_available(fd)
        readable, _, _ = select.select([fd], [], [], min(0.02, max(0.0, deadline - time.monotonic())))
        if readable:
            buffer += read_available(fd)
    return buffer


def write_bytes(fd: int, payload: bytes) -> None:
    offset = 0
    while offset < len(payload):
        offset += os.write(fd, payload[offset:])


def create_xclip(provider_dir: Path, payload_path: Path, log_path: Path) -> None:
    provider_dir.mkdir(parents=True, exist_ok=True)
    xclip = provider_dir / "xclip"
    xclip.write_text(
        "#!{python}\n"
        "import os\n"
        "import sys\n"
        "from pathlib import Path\n"
        "Path(os.environ['HADES_CLIPBOARD_LOG']).write_text(' '.join(sys.argv[1:]), encoding='utf-8')\n"
        "sys.stdout.buffer.write(Path(os.environ['HADES_CLIPBOARD_PAYLOAD']).read_bytes())\n".format(
            python=sys.executable
        ),
        encoding="utf-8",
    )
    xclip.chmod(0o755)
    payload_path.touch()
    log_path.touch()


def spawn(binary: Path, home: Path, provider_dir: Path, payload_path: Path, log_path: Path) -> tuple[int, int]:
    pid, fd = pty.fork()
    if pid == 0:
        environment = os.environ.copy()
        environment.update(
            {
                "TERM": "xterm-256color",
                "HERMES_HOME": str(home),
                "HOME": str(home),
                "PATH": f"{provider_dir}{os.pathsep}{environment.get('PATH', '')}",
                "HADES_CLIPBOARD_PAYLOAD": str(payload_path),
                "HADES_CLIPBOARD_LOG": str(log_path),
            }
        )
        for key in ("WAYLAND_DISPLAY", "WSL_INTEROP", "WSL_DISTRO_NAME"):
            environment.pop(key, None)
        os.execve(str(binary), [str(binary)], environment)

    set_window_size(fd)
    os.set_blocking(fd, False)
    return pid, fd


def stop_child(pid: int, fd: int) -> None:
    exited, _ = child_status(pid)
    if not exited:
        try:
            os.kill(pid, signal.SIGTERM)
        except ProcessLookupError:
            pass
        try:
            os.waitpid(pid, 0)
        except ChildProcessError:
            pass
    try:
        os.close(fd)
    except OSError:
        pass


def load_contract(path: Path) -> dict[str, Any]:
    try:
        contract = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise ClipboardReplayFailure("contract", "load", str(error)) from error
    if not isinstance(contract, dict) or contract.get("schema_version") != 1:
        raise ClipboardReplayFailure("contract", "load", "unsupported clipboard contract")
    steps = contract.get("steps")
    required = {"successful-text", "empty-provider-control"}
    ids = {step.get("id") for step in steps if isinstance(step, dict)} if isinstance(steps, list) else set()
    if ids != required:
        raise ClipboardReplayFailure("contract", "load", f"step ids must be {sorted(required)}")
    return contract


def input_bytes(event: dict[str, str]) -> bytes:
    try:
        return bytes.fromhex(event["bytes_hex"])
    except (KeyError, ValueError) as error:
        raise ClipboardReplayFailure("contract", "input", f"invalid input bytes: {error}") from error


def run_case(binary: Path, case: dict[str, Any], timeout: float, ordinal: int) -> dict[str, Any]:
    case_id = case["id"]
    home = Path(tempfile.mkdtemp(prefix=f"had023-{case_id}-{ordinal}-"))
    provider_dir = home / "bin"
    payload_path = home / "clipboard.payload"
    log_path = home / "clipboard.args"
    payload_path.write_bytes(case.get("provider_payload", "").encode("utf-8"))
    create_xclip(provider_dir, payload_path, log_path)
    pid: int | None = None
    fd: int | None = None
    buffer = b""
    try:
        pid, fd = spawn(binary, home, provider_dir, payload_path, log_path)
        startup_markers = ("Hermes Agent", "Nous Research", "Available Tools", "Available Skills")
        buffer = wait_for(
            pid,
            fd,
            buffer,
            case_id,
            "startup",
            lambda current: all(contains_marker(current, marker) for marker in startup_markers),
            timeout,
        )

        for event in case["input_sequence"]:
            write_bytes(fd, input_bytes(event))
            buffer = drain(pid, fd, buffer)

        expected = case["output"]
        markers = expected["screen_markers"]
        buffer = wait_for(
            pid,
            fd,
            buffer,
            case_id,
            "clipboard",
            lambda current: all(contains_marker(current, marker) for marker in markers),
            timeout,
        )
        for marker in expected.get("screen_absent_markers", []):
            if contains_marker(buffer, marker):
                raise ClipboardReplayFailure(case_id, "screen", f"unexpected marker: {marker}")

        buffer = wait_for(
            pid,
            fd,
            buffer,
            case_id,
            "provider",
            lambda _current: log_path.read_text(encoding="utf-8") == expected["provider_arguments"],
            timeout,
        )

        exited, status = child_status(pid)
        if exited:
            raise ClipboardReplayFailure(
                case_id,
                "ready-state",
                "process exited before cleanup",
                {"exit_status": status, "screen_tail": normalize(buffer)[-2000:]},
            )
        write_bytes(fd, b"\x03")
        buffer = wait_for(pid, fd, buffer, case_id, "exit", lambda _current: child_status(pid)[0], timeout)

        return {
            "id": case_id,
            "status": "passed",
            "input_bytes": [event["bytes_hex"] for event in case["input_sequence"]],
            "provider_arguments": expected["provider_arguments"],
            "observed": {
                "screen_markers": markers,
                "screen_absent_markers": expected.get("screen_absent_markers", []),
                "cleanup": "Ctrl+C exits from ready without submission",
            },
            "capture": "direct PTY with TIOCSWINSZ 120x40",
        }
    finally:
        if pid is not None and fd is not None:
            stop_child(pid, fd)
        shutil.rmtree(home, ignore_errors=True)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--binary", type=Path, default=DEFAULT_BINARY)
    parser.add_argument("--contract", type=Path, default=DEFAULT_CONTRACT)
    parser.add_argument("--report", type=Path, default=None)
    parser.add_argument("--timeout", type=float, default=5.0)
    arguments = parser.parse_args()
    binary = arguments.binary.resolve()
    contract_path = arguments.contract.resolve()
    report: dict[str, Any] = {
        "schema_version": 1,
        "command": "replay-clipboard-text",
        "passed": False,
        "binary": str(binary),
        "contract": str(contract_path),
        "checks": [],
    }

    try:
        if not binary.is_file():
            raise ClipboardReplayFailure("input", "binary", f"binary not found: {binary}")
        contract = load_contract(contract_path)
        report["contract_observation"] = contract["observation_id"]
        report["reference_observation"] = contract.get("reference_observation")
        report["dimensions"] = contract["reference"]["terminal"]
        for ordinal, case in enumerate(contract["steps"], start=1):
            report["checks"].append(run_case(binary, case, arguments.timeout, ordinal))
        report["passed"] = True
    except ClipboardReplayFailure as error:
        report["failure"] = error.as_dict()
    except (OSError, TypeError, ValueError, KeyError) as error:
        report["failure"] = {"case": "report", "step": "runtime", "message": str(error)}

    try:
        emit_report(report, arguments.report.resolve() if arguments.report else None)
    except OSError as error:
        print(json.dumps({"passed": False, "error": f"could not write report: {error}"}))
        return 2
    return 0 if report["passed"] else 1


if __name__ == "__main__":
    sys.exit(main())

File: crates/hades-dev/src/bin/replay_clipboard_text.rs
//! `hades-dev replay-clipboard-text` — Rust port of
//! `scripts/replay_clipboard_text.py` (HAD-147).
//!
//! Replays the Hades successful-text clipboard contract (OBS-0023) in an
//! isolated direct PTY with a synthetic xclip provider: seed text typed
//! into the composer, Ctrl+V (raw 0x16) invoking the fake xclip that
//! records its arguments and returns the fixture payload, the pasted
//! markers rendered on the reconstructed screen, the provider-argument
//! log matching the contract, and a clean Ctrl+C exit from ready.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::ExitCode;
use std::time::{Duration, Instant};

use hades_dev::pty::read_available;
use hades_dev::replay::{
    ExitStatus, ReplayChild, RetainedSlave, marker_present, spawn_with_env, terminal_flags,
    try_wait, wait_for, wait_for_exit, wait_for_rendered,
};
use hades_dev::screen::Screen;
use serde_json::{Value, json};

const STARTUP_MARKERS: [&str; 4] =
    ["Hades Agent", "Underworld", "Available Tools", "Available Skills"];
const STRIP_ENV: [&str; 3] = ["WAYLAND_DISPLAY", "WSL_INTEROP", "WSL_DISTRO_NAME"];

/// Reconstruct the on-screen text (what a real terminal shows). The
/// animated startup logo emits interleaved sparse-redraw cell writes that
/// fragment typed text in the raw stream; the Screen emulator rebuilds the
/// grid so draft/paste markers are contiguous again.
fn rendered_text(output: &[u8]) -> String {
    let mut screen = Screen::new(120, 40);
    screen.feed(output);
    screen.lines().join("\n")
}

struct Contract {
    observation_id: String,
    reference_observation: Option<String>,
    steps: Vec<Value>,
}

fn load_contract(path: &Path) -> Result<Contract, String> {
    let contents =
        fs::read_to_string(path).map_err(|error| format!("could not read contract: {error}"))?;
    let contract: Value =
        serde_json::from_str(&contents).map_err(|error| format!("invalid contract: {error}"))?;
    if contract.get("schema_version").and_then(Value::as_u64) != Some(1) {
        return Err("unsupported clipboard contract".to_owned());
    }
    let steps = contract
        .get("steps")
        .and_then(Value::as_array)
        .cloned()
        .ok_or_else(|| "contract has no steps".to_owned())?;
    let ids: Vec<String> = steps
        .iter()
        .filter_map(|step| step.get("id").and_then(Value::as_str).map(str::to_owned))
        .collect();
    let required = ["successful-text", "empty-provider-control"];
    let mut sorted = ids.clone();
    sorted.sort();
    let mut required_sorted = required.to_vec();
    required_sorted.sort();
    if sorted != required_sorted {
        return Err(format!("step ids must be {:?}", required.to_vec()));
    }
    Ok(Contract {
        observation_id: contract
            .get("observation_id")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned(),
        reference_observation: contract
            .get("reference_observation")
            .and_then(Value::as_str)
            .map(str::to_owned),
        steps,
    })
}

fn input_bytes(event: &Value) -> Result<Vec<u8>, String> {
    let hex = event
        .get("bytes_hex")
        .and_then(Value::as_str)
        .ok_or_else(|| "input event has no bytes_hex".to_owned())?;
    let compact: String = hex.split_whitespace().collect();
    let bytes = (0..compact.len())
        .step_by(2)
        .map(|index| {
            u8::from_str_radix(&compact[index..index + 2], 16)
                .map_err(|error| format!("invalid input bytes: {error}"))
        })
        .collect::<Result<Vec<_>, _>>()?;
    Ok(bytes)
}

/// Write the synthetic xclip provider: records argv into the log file and
/// echoes the fixture payload to stdout.
fn create_xclip(provider_dir: &Path, log_path: &Path) -> Result<(), String> {
    fs::create_dir_all(provider_dir).map_err(|error| error.to_string())?;
    let script =
        "#!/bin/sh\nprintf '%s\\n' \"$*\" > \"$HADES_CLIPBOARD_LOG\"\ncat \"$HADES_CLIPBOARD_PAYLOAD\"\n"
            .to_string();
    let xclip = provider_dir.join("xclip");
    fs::write(&xclip, script).map_err(|error| error.to_string())?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut permissions =
            fs::metadata(&xclip).map_err(|error| error.to_string())?.permissions();
        permissions.set_mode(0o755);
        fs::set_permissions(&xclip, permissions).map_err(|error| error.to_string())?;
    }
    // Note: unlike Python's `Path.touch()`, `fs::write` truncates — the
    // payload file must keep the fixture bytes already written by run_case,
    // so only the (initially empty) argument log is created here.
    fs::write(log_path, []).map_err(|error| error.to_string())?;
    Ok(())
}

fn spawn_with_provider(
    binary: &Path,
    home: &Path,
    provider_dir: &Path,
    payload_path: &Path,
    log_path: &Path,
) -> Result<ReplayChild, String> {
    let path_value = std::env::var("PATH").unwrap_or_default();
    // PATH entries are separated by the platform's path list separator
    // (':' on Unix) — use join_paths, not MAIN_SEPARATOR.
    let mut entries = vec![provider_dir.as_os_str().to_owned()];
    entries.extend(std::env::split_paths(&path_value).map(std::ffi::OsString::from));
    let provider_path =
        std::env::join_paths(entries).map_err(|error| format!("could not build PATH: {error}"))?;
    let provider_path =
        provider_path.to_str().ok_or_else(|| "non-utf8 provider path".to_owned())?.to_owned();
    let extra: [(&str, &str); 6] = [
        ("HOME", home.to_str().ok_or_else(|| "non-utf8 home".to_owned())?),
        ("HERMES_HOME", home.to_str().ok_or_else(|| "non-utf8 home".to_owned())?),
        ("HADES_PROVIDER_BASE_URL", "http://127.0.0.1:8765/v1"),
        ("PATH", &provider_path),
        (
            "HADES_CLIPBOARD_PAYLOAD",
            payload_path.to_str().ok_or_else(|| "non-utf8 payload".to_owned())?,
        ),
        ("HADES_CLIPBOARD_LOG", log_path.to_str().ok_or_else(|| "non-utf8 log".to_owned())?),
    ];
    spawn_with_env(binary, &[], &extra, &STRIP_ENV).map_err(|error| error.to_string())
}

fn run_case(
    binary: &Path,
    case: &Value,
    timeout: Duration,
    ordinal: usize,
) -> Result<Value, String> {
    let case_id = case
        .get("id")
        .and_then(Value::as_str)
        .ok_or_else(|| "case has no id".to_owned())?
        .to_owned();
    let home =
        std::env::temp_dir().join(format!("had023-{case_id}-{ordinal}-{}", std::process::id()));
    let provider_dir = home.join("bin");
    let payload_path = home.join("clipboard.payload");
    let log_path = home.join("clipboard.args");
    fs::create_dir_all(&home).map_err(|error| error.to_string())?;
    let payload = case
        .get("provider_payload")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .as_bytes()
        .to_vec();
    fs::write(&payload_path, &payload).map_err(|error| error.to_string())?;
    create_xclip(&provider_dir, &log_path)?;

    let mut reaped = false;
    let result = (|| -> Result<Value, String> {
        let child = spawn_with_provider(binary, &home, &provider_dir, &payload_path, &log_path)?;
        let mut output = Vec::new();
        let slave_path = child.slave_path.clone();

        wait_for(
            &child.child,
            &mut output,
            &format!("{case_id}: startup"),
            |text| STARTUP_MARKERS.iter().all(|marker| marker_present(text, marker)),
            timeout,
        )?;

        for event in
            case.get("input_sequence").and_then(Value::as_array).cloned().unwrap_or_default()
        {
            hades_dev::replay::send(&child.child.master, &input_bytes(&event)?)
                .map_err(|error| error.to_string())?;
            std::thread::sleep(Duration::from_millis(80));
            output.extend_from_slice(&read_available(&child.child.master));
        }

        let expected =
            case.get("output").ok_or_else(|| "case has no output contract".to_owned())?.clone();
        let markers: Vec<String> = expected
            .get("screen_markers")
            .and_then(Value::as_array)
            .map(|markers| {
                markers.iter().filter_map(|marker| marker.as_str().map(str::to_owned)).collect()
            })
            .unwrap_or_default();
        wait_for_rendered(
            &child.child,
            &mut output,
            &format!("{case_id}: clipboard"),
            |rendered| markers.iter().all(|marker| marker_present(rendered, marker)),
            timeout,
        )?;

        let rendered = rendered_text(&output);
        for marker in expected
            .get("screen_absent_markers")
            .and_then(Value::as_array)
            .map(|markers| {
                markers
                    .iter()
                    .filter_map(|marker| marker.as_str().map(str::to_owned))
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default()
        {
            if marker_present(&rendered, &marker) {
                return Err(format!("{case_id}: unexpected marker: {marker}"));
            }
        }

        let provider_arguments = expected
            .get("provider_arguments")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned();
        let log_deadline = Instant::now() + timeout;
        loop {
            let log_contents = fs::read_to_string(&log_path).unwrap_or_default();
            if log_contents.trim_end() == provider_arguments {
                break;
            }
            if try_wait(&child.child)?.is_some() {
                return Err(format!("{case_id}: provider: process exited before xclip log match"));
            }
            if Instant::now() >= log_deadline {
                return Err(format!("{case_id}: provider: timed out waiting for xclip arguments"));
            }
            std::thread::sleep(Duration::from_millis(50));
        }

        if try_wait(&child.child)?.is_some() {
            return Err(format!("{case_id}: process exited before cleanup"));
        }
        let slave = RetainedSlave::retain(&slave_path).map_err(|error| error.to_string())?;
        let ready_flags = terminal_flags(&slave).map_err(|error| error.to_string())?;
        if ready_flags.canonical || ready_flags.echo {
            return Err(format!("{case_id}: ready state was not raw mode: {ready_flags:?}"));
        }

        hades_dev::replay::send(&child.child.master, b"\x03").map_err(|error| error.to_string())?;
        let status = wait_for_exit(&child.child, &mut output, timeout)?;
        reaped = true;
        let exit_status = ExitStatus::describe(status);
        if exit_status != (ExitStatus::Exit { code: 0 }) {
            return Err(format!("{case_id}: unexpected exit status: {exit_status:?}"));
        }

        Ok(json!({
            "id": case_id,
            "status": "passed",
            "input_bytes": case
                .get("input_sequence")
                .and_then(Value::as_array)
                .map(|events| {
                    events
                        .iter()
                        .filter_map(|event| event.get("bytes_hex").and_then(Value::as_str))
                        .collect::<Vec<_>>()
                })
                .unwrap_or_default(),
            "provider_arguments": provider_arguments,
            "observed": {
                "screen_markers": markers,
                "screen_absent_markers": expected
                    .get("screen_absent_markers")
                    .and_then(Value::as_array)
                    .cloned()
                    .unwrap_or_default(),
                "cleanup": "Ctrl+C exits from ready without submission",
            },
            "capture": "direct PTY with TIOCSWINSZ 120x40",
        }))
    })();

    if !reaped {
        let _ = fs::remove_dir_all(&home);
    }
    result
}

fn write_report(report: &Value, path: Option<&Path>, status: u8) -> ExitCode {
    let text = serde_json::to_string_pretty(report).expect("serialize report");
    if let Some(path) = path {
        if let Some(parent) = path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        use std::io::Write;
        if let Ok(mut file) = fs::File::create(path) {
            let _ = file.write_all(text.as_bytes());
            let _ = file.write_all(b"\n");
        }
    }
    println!("{text}");
    if status == 0 { ExitCode::SUCCESS } else { ExitCode::from(status) }
}

fn main() -> ExitCode {
    let mut args = std::env::args().skip(1);
    let mut binary = PathBuf::from("target/debug/hades");
    let mut contract_path =
        PathBuf::from("tests/fixtures/parity/OBS-0023-hades-text-clipboard.json");
    let mut report_path: Option<PathBuf> = None;
    let mut timeout = Duration::from_secs_f64(5.0);

    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--binary" => {
                if let Some(value) = args.next() {
                    binary = PathBuf::from(value);
                }
            }
            "--contract" => {
                if let Some(value) = args.next() {
                    contract_path = PathBuf::from(value);
                }
            }
            "--report" => {
                if let Some(value) = args.next() {
                    report_path = Some(PathBuf::from(value));
                }
            }
            "--timeout" => {
                if let Some(value) = args.next() {
                    timeout = Duration::from_secs_f64(value.parse().unwrap_or(5.0));
                }
            }
            _ => {}
        }
    }

    let binary = binary.canonicalize().unwrap_or_else(|_| binary.clone());
    let contract_path = contract_path.canonicalize().unwrap_or_else(|_| contract_path.clone());
    let mut report = json!({
        "schema_version": 1,
        "command": "replay-clipboard-text",
        "passed": false,
        "binary": binary.display().to_string(),
        "contract": contract_path.display().to_string(),
        "checks": [],
    });
    if !binary.is_file() {
        report["failure"] = json!({"case": "input", "step": "binary", "message": format!("binary not found: {}", binary.display())});
        return write_report(&report, report_path.as_deref(), 1);
    }

    match load_contract(&contract_path) {
        Ok(contract) => {
            report["contract_observation"] = json!(contract.observation_id);
            if let Some(reference) = contract.reference_observation {
                report["reference_observation"] = json!(reference);
            }
            report["dimensions"] = json!({
                "columns": 120,
                "rows": 40,
                "emulator": "direct PTY with raw byte writes",
            });
            let mut checks = Vec::new();
            for (ordinal, case) in contract.steps.iter().enumerate() {
                match run_case(&binary, case, timeout, ordinal + 1) {
                    Ok(check) => checks.push(check),
                    Err(error) => {
                        report["failure"] =
                            json!({"case": "report", "step": "runtime", "message": error});
                        return write_report(&report, report_path.as_deref(), 1);
                    }
                }
            }
            report["checks"] = json!(checks);
            report["passed"] = json!(true);
            write_report(&report, report_path.as_deref(), 0)
        }
        Err(error) => {
            report["failure"] = json!({"case": "contract", "step": "load", "message": error});
            write_report(&report, report_path.as_deref(), 1)
        }
    }
}
